File: android/src/main/java/org/ligi/gobandroid_hd/ui/recording/SaveSGFDialog.java
package org.ligi.gobandroid_hd.ui.recording;

import android.content.DialogInterface;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.CompoundButton.OnCheckedChangeListener;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import org.ligi.gobandroid_hd.App;
import org.ligi.gobandroid_hd.R;
import org.ligi.gobandroid_hd.logic.GoGameMetadata;
import org.ligi.gobandroid_hd.logic.sgf.SGFWriter;
import org.ligi.gobandroid_hd.ui.GobandroidDialog;
import org.ligi.gobandroid_hd.ui.application.GobandroidFragmentActivity;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.Date;

/**
 * Dialog to save a game to SGF file and ask the user about how in here
 * <p/>
 * TODO check if file exists
 *
 * @author ligi
 */
public class SaveSGFDialog extends GobandroidDialog {

    private EditText fname_et;
    private GobandroidFragmentActivity context;
    private CheckBox override_checkbox;

    public SaveSGFDialog(GobandroidFragmentActivity _context) {
        super(_context);

        context = _context;

        setContentView(R.layout.dialog_save_sgf);

        setIconResource(R.drawable.save);
        TextView intro_text = (TextView) findViewById(R.id.intro_txt);

        override_checkbox = (CheckBox) findViewById(R.id.override_checkbox);

        intro_text.setText(String.format(context.getResources().getString(R.string.save_sgf_question), context.getSettings().getSGFSavePath()));

        fname_et = (EditText) findViewById(R.id.sgf_name_edittext);

        class SaveSGFOnClickListener implements DialogInterface.OnClickListener {
            public void onClick(DialogInterface dialog, int whichButton) {
                String fname = getCompleteFileName();
                boolean res = SGFWriter.saveSGF(App.getGame(), fname);

                if (res)
                    Toast.makeText(context, String.format(context.getString(R.string.file_saved), fname), Toast.LENGTH_SHORT).show();
                else
                    Toast.makeText(context, String.format(context.getString(R.string.file_not_saved), fname), Toast.LENGTH_SHORT).show();

                dialog.dismiss();
            }

        }

        setPositiveButton(android.R.string.ok, new SaveSGFOnClickListener());

        fname_et.addTextChangedListener(new TextWatcher() {

            @Override
            public void afterTextChanged(Editable s) {
                setPositiveButtonAndOverrideCheckboxEnabledByExistenceOfFile();
            }

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

        });

        override_checkbox.setOnCheckedChangeListener(new OnCheckedChangeListener() {

            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                setPositiveButtonAndOverrideCheckboxEnabledByExistenceOfFile();

            }

        });

        // get the old filename from the metadata
        String old_fname = App.getGame().getMetaData().getFileName();

        if ((old_fname != null) && (!old_fname.equals(""))) {
            String suggested_name = old_fname.replace(".sgf", "");
            if (suggested_name.startsWith(context.getSettings().getSGFSavePath()))
                suggested_name = suggested_name.substring(context.getSettings().getSGFSavePath().length());
            fname_et.setText(suggested_name);
        }

        final GoGameMetadata game_meta = App.getGame().getMetaData();

        /**
         * this is a OnClickListener to add Stuff to the FileName like
         * date/gamename/...
         */
        class FileNameAdder implements View.OnClickListener {

            private String getTextByButtonId(int btn_resId) {
                switch (btn_resId) {
                    case R.id.button_add_date:
                        SimpleDateFormat date_formatter = new SimpleDateFormat("yyyy.MM.dd");
                        return date_formatter.format(new Date());
                    case R.id.button_add_time:
                        SimpleDateFormat time_formatter = new SimpleDateFormat("H'h'm'm'");
                        return time_formatter.format(new Date());
                    case R.id.button_add_gamename:
                        return game_meta.getName();
                    case R.id.button_add_players:
                        return game_meta.getBlackName() + "_vs_" + game_meta.getWhiteName();

                    default:
                        return null;
                }
            }

            @Override
            public void onClick(View v) {
                String toAdd = getTextByButtonId(v.getId());
                if (toAdd != null) {
                    String text = fname_et.getText().toString();
                    int cursorPos = fname_et.getSelectionStart();
                    StringBuilder sb = new StringBuilder();
                    sb.append(text.substring(0, cursorPos)).append(toAdd).append(text.substring(cursorPos, fname_et.length()));
                    fname_et.setText(sb.toString());
                    fname_et.setSelection(cursorPos + toAdd.length());
                }
            }

        }
        FileNameAdder adder = new FileNameAdder();

        ((Button) (findViewById(R.id.button_add_date))).setOnClickListener(adder);
        ((Button) (findViewById(R.id.button_add_time))).setOnClickListener(adder);
        Button add_name_btn = ((Button) (findViewById(R.id.button_add_gamename)));
        Button players_name_btn = ((Button) (findViewById(R.id.button_add_players)));

        if (game_meta.getName().equals(""))
            add_name_btn.setVisibility(View.GONE);
        else
            add_name_btn.setOnClickListener(adder);

        if (game_meta.getBlackName().equals("") && game_meta.getWhiteName().equals(""))
            players_name_btn.setVisibility(View.GONE);
        else
            players_name_btn.setOnClickListener(adder);

        setTitle(R.string.save_sgf);

        setPositiveButtonAndOverrideCheckboxEnabledByExistenceOfFile();

    }

    private void setPositiveButtonAndOverrideCheckboxEnabledByExistenceOfFile() {

        String fname = getCompleteFileName();

        if (fname == null) { // we got no filename from user

            override_checkbox.setVisibility(View.GONE); // no overwrite without
            // filename
            getPositiveButton().setEnabled(false); // should not save without a
            // filename
            return;
        }

        File wanted_file = new File(getCompleteFileName());
        boolean target_file_exist = wanted_file.exists();
        override_checkbox.setVisibility((target_file_exist && !wanted_file.isDirectory()) ? View.VISIBLE : View.GONE);
        getPositiveButton().setEnabled(!target_file_exist || override_checkbox.isChecked());
    }

    /**
     * @return the filename with path and file extension - returns null when
     * there is no filename given
     */
    private String getCompleteFileName() {
        String fname = fname_et.getText().toString();

        if (fname.length() == 0)
            return null;

        fname += ".sgf"; // append filename extension

        if (fname.startsWith("/"))
            return fname;
        else
            return context.getSettings().getSGFSavePath() + fname;
    }

}

File: android/src/main/java/org/ligi/gobandroid_hd/ui/recording/SaveSGFDialog.kt
package org.ligi.gobandroid_hd.ui.recording

import android.view.View
import android.widget.Button
import android.widget.Toast
import kotlinx.android.synthetic.main.dialog_save_sgf.*
import org.ligi.gobandroid_hd.R
import org.ligi.gobandroid_hd.logic.sgf.SGFWriter
import org.ligi.gobandroid_hd.ui.GobandroidDialog
import org.ligi.gobandroid_hd.ui.application.GobandroidFragmentActivity
import org.ligi.kaxt.doAfterEdit
import org.ligi.kaxt.setVisibility
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

/**
 * Dialog to save a game to SGF file and ask the user about how in here
 *
 *
 * TODO check if file exists

 * @author ligi
 */
class SaveSGFDialog(private val context: GobandroidFragmentActivity) : GobandroidDialog(context) {

    init {

        setContentView(R.layout.dialog_save_sgf)

        setIconResource(R.drawable.ic_content_save)

        intro_txt.text = String.format(context.resources.getString(R.string.save_sgf_question), settings.SGFSavePath)

        setPositiveButton(android.R.string.ok, { _ ->
            val file = completeFileName()
            val res = SGFWriter.saveSGF(gameProvider.get(), file!!)

            val toastText = if (res) R.string.file_saved else R.string.file_not_saved
            Toast.makeText(context, String.format(context.getString(toastText), file.absolutePath), Toast.LENGTH_SHORT).show()
        })

        sgf_name_edittext.doAfterEdit {
            setPositiveButtonAndOverrideCheckboxEnabledByExistenceOfFile()
        }

        override_checkbox.setOnCheckedChangeListener { _, _ -> setPositiveButtonAndOverrideCheckboxEnabledByExistenceOfFile() }

        // get the old filename from the metadata
        val oldFileName = gameProvider.get().metaData.fileName

        if (oldFileName.isNotBlank()) {
            var suggested_name = oldFileName.replace(".sgf", "")
            val absolutePath = settings.SGFSavePath.absolutePath
            if (suggested_name.startsWith(absolutePath)) {
                suggested_name = suggested_name.substring(absolutePath.length + 1)
            }
            sgf_name_edittext.setText(suggested_name)
        }

        val (name, _, _, _, blackName, _, whiteName) = gameProvider.get().metaData

        /**
         * this is a OnClickListener to add Stuff to the FileName like
         * date/gamename/...
         */
        class FileNameAdder : View.OnClickListener {

            private fun getTextByButtonId(btn_resId: Int) = when (btn_resId) {
                R.id.button_add_date -> {
                    val date_formatter = SimpleDateFormat("yyyy.MM.dd")
                    date_formatter.format(Date())
                }
                R.id.button_add_time -> {
                    val time_formatter = SimpleDateFormat("H'h'm'm'")
                    time_formatter.format(Date())
                }
                R.id.button_add_gamename -> name
                R.id.button_add_players -> blackName + "_vs_" + whiteName

                else -> null
            }


            override fun onClick(v: View) {
                val toAdd = getTextByButtonId(v.id)
                if (toAdd != null) {
                    val text = sgf_name_edittext.text.toString()
                    val cursorPos = sgf_name_edittext.selectionStart
                    val sb = StringBuilder()
                    sb.append(text.substring(0, cursorPos)).append(toAdd).append(text.substring(cursorPos, sgf_name_edittext.length()))
                    sgf_name_edittext.setText(sb.toString())
                    sgf_name_edittext.setSelection(cursorPos + toAdd.length)
                }
            }

        }

        val adder = FileNameAdder()

        button_add_date.setOnClickListener(adder)
        button_add_time.setOnClickListener(adder)

        button_add_gamename.prepareButton(name.isBlank(), adder)
        button_add_players.prepareButton(blackName.isBlank() && whiteName.isBlank(), adder)

        setTitle(R.string.save_sgf)

        setPositiveButtonAndOverrideCheckboxEnabledByExistenceOfFile()
    }

    fun Button.prepareButton(condition: Boolean, adder: View.OnClickListener) = this.apply {
        setVisibility(condition)
        setOnClickListener(adder)
    }

    private fun setPositiveButtonAndOverrideCheckboxEnabledByExistenceOfFile() {

        val wanted_file = completeFileName()

        if (wanted_file == null) { // we got no filename from user

            override_checkbox.visibility = View.GONE // no overwrite without
            // filename

            positive_btn.isEnabled = false // should not save without a
            // filename
            return
        }

        val target_file_exist = wanted_file.exists()
        override_checkbox.visibility = if (target_file_exist && !wanted_file.isDirectory) View.VISIBLE else View.GONE
        positive_btn.isEnabled = !target_file_exist || override_checkbox.isChecked
    }

    /**
     * @return the filename with path and file extension
     * or null when there is no filename given
     */
    private fun completeFileName(): File? {
        var fileName = sgf_name_edittext.text.toString()

        if (fileName.isEmpty())
            return null

        fileName += ".sgf"

        return if (fileName.startsWith("/"))
            File(fileName)
        else
            File(settings.SGFSavePath, fileName)
    }

}
